libvmaf/src/feature/integer_vif.c

```diff
@@ -37,758 +53,1033 @@
+			int32_t sigma2_sq = (int32_t)(yy_filt_val - mu2_sq_val);
+			int32_t sigma12 = (int32_t)(xy_filt_val - mu1_mu2_val);
+
+			sigma2_sq = MAX(sigma2_sq, 0);
+			if (sigma1_sq >= sigma_nsq) {
+				uint32_t log_den_stage1 = (uint32_t)(sigma_nsq + sigma1_sq);
+				int x;
+				uint16_t log_den1 = get_best16_from32(log_den_stage1, &x);
+
+				/**
+				* log values are taken from the look-up table generated by
+				* log_generate() function which is called in integer_combo_threadfunc
+				* den_val in float is log2(1 + sigma1_sq/2)
+				* here it is converted to equivalent of log2(2+sigma1_sq) - log2(2) i.e log2(2*65536+sigma1_sq) - 17
+				* multiplied by 2048 as log_value = log2(i)*2048 i=16384 to 65535 generated using log_value
+				* x because best 16 bits are taken
+				*/
+				accum_x += x + 17;
+				den_val = log2_table[log_den1];
+
+				// this check can go away, things will work anyway
+				if (sigma12 > 0) {
+					// num_val = log2f(1.0f + (g * g * sigma1_sq) / (sv_sq + sigma_nsq));
+					/**
+					* In floating-point numerator = log2((1.0f + (g * g * sigma1_sq)/(sv_sq + sigma_nsq))
+					*
+					* In Fixed-point the above is converted to
+					* numerator = log2((sv_sq + sigma_nsq)+(g * g * sigma1_sq))- log2(sv_sq + sigma_nsq)
+					*/
+
+					const double eps = 65536 * 1.0e-10;
+					double g = sigma12 / (sigma1_sq + eps); // this epsilon can go away
+					int32_t sv_sq = sigma2_sq - g * sigma12;
+
+					if (sigma2_sq == 0) { // this was... sigma2_sq < eps
+						g = 0.0;
+					}
+
+					sv_sq = (uint32_t)(MAX(sv_sq, 0));
+
+					g = MIN(g, vif_enhn_gain_limit);
+
+					int x1, x2;
+					uint32_t numer1 = (sv_sq + sigma_nsq);
+					int64_t numer1_tmp = (int64_t)((g * g * sigma1_sq)) + numer1; //numerator
+					uint16_t numlog = get_best16_from64((uint64_t)numer1_tmp, &x1);
+					uint16_t denlog = get_best16_from64((uint64_t)numer1, &x2);
+					accum_x2 += (x2 - x1);
+					num_val = log2_table[numlog] - log2_table[denlog];
+					accum_num_log += num_val;
+					accum_den_log += den_val;
+
+				}
+				else {
+					//accum_num_log += 0;
+					accum_den_log += den_val;
+				}
+			}
+			else {
+				accum_num_non_log += sigma2_sq;
+				accum_den_non_log++;
+			}
+		}
+	}
+	num[0] = accum_num_log / 2048.0 + accum_x2 + (accum_den_non_log - ((accum_num_non_log) / 16384.0) / (65025.0));
+	den[0] = accum_den_log / 2048.0 - accum_x + accum_den_non_log;
+}
+
+void vif_statistic_16(struct VifState* s, float* num, float* den, unsigned w, unsigned h, int bpc, int scale) {
+	const unsigned fwidth = vif_filter1d_width[scale];
+	const uint16_t* vif_filt = vif_filter1d_table[scale];
+	VifBuffer buf = s->buf;
+	int64_t num_val, den_val;
+	int64_t accum_x = 0, accum_x2 = 0;
+	int64_t accum_num_log = 0.0;
+	int64_t accum_den_log = 0.0;
+	int64_t accum_num_non_log = 0;
+	int64_t accum_den_non_log = 0;
+	uint32_t* xx_filt = buf.ref_sq;
+	uint32_t* yy_filt = buf.dis_sq;
+	uint32_t* xy_filt = buf.ref_dis;
+	static const int32_t sigma_nsq = 65536 << 1;
```

Comment:
Why is it identical to sigma_nsq from vif_statistic_8? For 10 bit it should be 32 (2 * 4^2), for 12 bit - 512 (2 * 16^2).

Reply:
Hi igv,
really thank you for pointing out.
I did not consider 10/12 bit at all, and this must simply be a bug.
I'm really sorry for the hurried pull request, but for me this is just a proof of concept.
Probably "pull request", "proof of concept", and "regression tests broken" are not good friends.
I'm very new to open source contributions and collaboration with unknown people.
I'll fix that shortly.

Reply:
Hi igv,
I looked at integer_vif implementation, and I really can't get the point:

This is the original implementation, and... can't see anything related to 10 or 12 bits.

```
static void vif_statistic(VifBuffer buf, float *num, float *den,
                          unsigned w, unsigned h, uint16_t *log2_table,
                          double vif_enhn_gain_limit)
{
    uint32_t *xx_filt = buf.ref_sq;
    uint32_t *yy_filt = buf.dis_sq;
    uint32_t *xy_filt = buf.ref_dis;

    //float equivalent of 2. (2 * 65536)
    static const int32_t sigma_nsq = 65536 << 1;
...
```
 
Can you help me understand?

Reply:
Was it even compatible with higher bit depths? Original VIF implementation certainly isn't. I'm assuming `vif_statistic_16` means for 16 bit inputs (10-16), right?

Reply:
Hi igv,
vif_statistic_16 is actually both for > 8 bit and for second stage and so on.

I really did not spend enough time with > 8 bit implementation, I'm still at an early stage.
So... there might be bugs, I was superclear about that! 
Let's see what the regression tests say.

But... what do you mean by "Original VIF implementation certainly isn't"? You mean that the head of master branch is not ok?

Reply:
I'm not very familiar with VMAF code base, just trying to make sure that this new version of VIF is optimally optimized and without bugs.
That sigma_nsq constant (variance of invisible noise) came from matlab version of VIFp, which is for 8bit images only. If your `vif_statistic_16` actually processes >=10bit frames (AFAIK currently VMAF converts everything to 8bit internally) than you need to increase that constant properly (and optionally epsilon).